### PR TITLE
0.2.26

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: src/contract.ts [hunk 1]
```diff
@@ -111,25 +113,35 @@ export class AbstractContract {
...
     // some time there is no opcodes, such as when sourcemap flag is closeed. 
     if(opcodes) {
       const offset = unlockingScriptASM.trim().split(' ').length;
       // the complete script may have op_return and data, but compiled output does not have it. So we need to make sure the index is in boundary.
-      const pc = Math.min(stepCounter -  offset,  opcodes.length -1);
-  
+
+      const lastStepIndex = AbstractContract.findLastfExec(steps, stepCounter);
+
+      if(this._dataPart !== undefined && this._dataPart !== null) {
```
Comment (xhliu) [inline src/contract.ts:125]: simplify
`if(this._dataPart != null) {`

File: src/contract.ts [hunk 2]
```diff
@@ -111,25 +113,35 @@ export class AbstractContract {
...
+        opcodes.push({opcode: 'OP_RETURN', file: undefined, line: undefined, endLine: undefined, column: undefined, endColumn: undefined, stack:[]});
```
Comment (xhliu) [inline src/contract.ts:126]: This does not look good, maybe refactor like below
```
export interface OpCode {
	file?: string;
	line?: number;
	endLine?: number;
	column?: number;
	endColumn: number;
	opcode: string;
	stack: string[];
	debugTag?: DebugModeTag;
}
```
->
```
export interface Pos {
	file?: string;
	line?: number;
	endLine?: number;
	column?: number;
	endColumn: number;
}
export interface OpCode {
	opcode: string;
	stack: string[];
         pos?: Pos;
	debugTag?: DebugModeTag;
}
```

File: src/contract.ts [hunk 3]
```diff
@@ -111,25 +113,35 @@ export class AbstractContract {
...
+        this._dataPart.split(' ').forEach(data => {
+          opcodes.push({opcode: data, file: undefined, line: undefined, endLine: undefined, column: undefined, endColumn: undefined, stack:[]});
+        });
+      }
+      let pc = lastStepIndex -  offset;
+      if(stepCounter === (opcodes.length + offset)) { // all opcode  was exec
+        
+      } else { //not all opcode was exec, break like OP_VERIFY
+        pc += 1;
+      }
```
Comment (xhliu) [inline src/contract.ts:136]:
```
      if(stepCounter < (opcodes.length + offset)) {
       // not all opcodes were executed, stopped in the middle at opcode like OP_VERIFY
        pc += 1;
      }
```

File: out/ackermann_desc.json [hunk 4]
```diff
@@ -31,7 +31,7 @@
...
     "sources": [
-        "D:\\code\\scryptlib\\test\\fixture\\ackermann.scrypt"
+        "C:\\Users\\myland\\code\\scryptlib\\test\\fixture\\ackermann.scrypt"
```
Comment (xhliu) [inline out/ackermann_desc.json:34]: Why did this change? Let's try minimizing unnecessary changes from now on, if not this time.

**Review by xhliu — CHANGES_REQUESTED** [review]
Looking good 🥇 

Why `tokenUtxo` contract? If u only wanna test contract w/ dependencies, just use bar.scrypt. Library code does not need to test complicate contract, which is unnecessary & hard to maintain.

File: out/ackermann_desc.json (hunk 4 above)
Comment (zhfnjust) [reply out/ackermann_desc.json:34]: previeus commit is by my book, now is by my pc. so path change.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,13 @@
 # CHANGELOG
 
+## 0.2.26
+
+* add gotosource test
+* Fix failed when launch debugger from test error
+* Fix source map failed on boilerplate/tests/js/tokenUtxo.scrypttest.js
```
[the remark this answers is not in the document]
Comment (zhfnjust) [reply CHANGELOG.md]: done

File: src/contract.ts (hunk 1 above)
Comment (zhfnjust) [reply src/contract.ts]: done

File: src/contract.ts (hunk 2 above)
Comment (zhfnjust) [reply src/contract.ts]: done

File: src/contract.ts (hunk 3 above)
Comment (zhfnjust) [reply src/contract.ts]: done

File: out/ackermann_desc.json (hunk 4 above)
Comment (zhfnjust) [reply out/ackermann_desc.json:36]: no change

File: out/simple_desc.json [hunk 6]
```diff
@@ -1,4 +1,5 @@
 {
+    "version": 1,
```
Comment (freedomhero) [inline out/simple_desc.json:2]: Maybe use a string as version to make it more readable, for example "20201226"

File: src/compilerWrapper.ts [hunk 7]
```diff
@@ -13,7 +13,7 @@ const SEMANTIC_ERR_REG = /Error:\s*(?<filePath>[^\s]+):(?<line>\d+):(?<column>\d
 const IMPORT_ERR_REG = /Syntax error:\s*(?<filePath>[^\s]+):(?<line>\d+):(?<column>\d+):\n([^\n]+\n){3}File not found: (?<fileName>[^\s]+)/g;
 //SOURCE_REG parser src eg: [0:6:3:8:4#Bar.constructor:0]
 const SOURCE_REG =  /^(?<fileIndex>-?\d+):(?<line>\d+):(?<col>\d+):(?<endLine>\d+):(?<endCol>\d+)(#(?<tagStr>.+))?/;
-
+const CURRENT_CONTRACTDESCRIPTION_VERSION = 1;
```
Comment (freedomhero) [inline src/compilerWrapper.ts:16]: -> CURRENT_CONTRACT_DESCRIPTION_VERSION

File: src/contract.ts [hunk 8]
```diff
@@ -46,7 +49,7 @@ export class AbstractContract {
 
   get lockingScript(): Script {
     let lsASM = this.scriptedConstructor.toASM();
-    if (this._dataPart !== undefined && this._dataPart !== null) {
+    if (this._dataPart) {
```
Comment (freedomhero) [inline src/contract.ts:52]: this._dataPart could be empty string, in which case, `opreturn` should also be added.

Comment (freedomhero) [reply src/contract.ts:52]: Could change to:
if (this._dataPart) {
   lsASM += ` OP_RETURN ${this._dataPart}`;
} else if (this._dataPart.trim() === '') {
  lsASM += ` OP_RETURN`;
}

File: src/contract.ts [hunk 9]
```diff
@@ -111,37 +114,45 @@ export class AbstractContract {
...
       // the complete script may have op_return and data, but compiled output does not have it. So we need to make sure the index is in boundary.
-      const pc = Math.min(stepCounter -  offset,  opcodes.length -1);
-  
-      if(!result && opcodes[pc]) {
-  
-        const opcode = opcodes[pc]; 
+
+      const lastStepIndex = AbstractContract.findLastfExec(steps, stepCounter);
+
+      if(this._dataPart) {
```
Comment (freedomhero) [inline src/contract.ts:126]: Same as above，this._dataPart could be empty string.

File: out/simple_desc.json (hunk 6 above)
Comment (zhfnjust) [reply out/simple_desc.json:2]: I think a number make compare logic more clear.

File: src/compilerWrapper.ts (hunk 7 above)
Comment (zhfnjust) [reply src/compilerWrapper.ts:16]: Fix

File: src/contract.ts (hunk 8 above)
Comment (zhfnjust) [reply src/contract.ts:52]: if (typeof this._dataPart === 'string') {
      const dp = this._dataPart.trim();
      if (dp) {
        lsASM += ` OP_RETURN ${dp}`;
      } else {
        lsASM += ` OP_RETURN`;  // note there is no space after op_return
      }
    }

File: src/contract.ts (hunk 9 above)
Comment (zhfnjust) [reply src/contract.ts]: f6181c8e687b92b652820e621b113ab6837ec23c

File: src/contract.ts (hunk 8 above)
Comment (zhfnjust) [reply src/contract.ts]: f6181c8e687b92b652820e621b113ab6837ec23c

File: src/contract.ts (hunk 2 above)
Comment (zhfnjust) [reply src/contract.ts]: 4e6853c0d45ee3524a7c726ae3fca1a984d9d3c9
